File: apps/erpnext_fiscal_br/erpnext_fiscal_br/events/item.py
"""
Eventos para Item
"""

import frappe
from frappe import _


def validate(doc, method):
    """
    Valida dados fiscais do item
    """
    try:
        from erpnext_fiscal_br.utils.cnpj_cpf import validar_ncm, validar_cest
    except ImportError:
        return
    
    # Valida NCM
    if doc.get("ncm"):
        ncm = str(doc.ncm).replace(".", "").replace("-", "")
        
        if not validar_ncm(ncm):
            frappe.throw(_("NCM deve ter 8 dígitos: {0}").format(doc.ncm))
        
        doc.ncm = ncm
        
        # Preenche gênero automaticamente
        if not doc.get("genero"):
            doc.genero = ncm[:2]
    
    # Valida CEST
    if doc.get("cest"):
        cest = str(doc.cest).replace(".", "").replace("-", "")
        
        if not validar_cest(cest):
            frappe.throw(_("CEST deve ter 7 dígitos: {0}").format(doc.cest))
        
        doc.cest = cest
    
    # Valida CFOP
    if doc.get("cfop_venda_interna"):
        cfop = str(doc.cfop_venda_interna).replace(".", "")
        if len(cfop) != 4:
            frappe.throw(_("CFOP de venda interna deve ter 4 dígitos"))
        doc.cfop_venda_interna = cfop
    
    if doc.get("cfop_venda_interestadual"):
        cfop = str(doc.cfop_venda_interestadual).replace(".", "")
        if len(cfop) != 4:
            frappe.throw(_("CFOP de venda interestadual deve ter 4 dígitos"))
        doc.cfop_venda_interestadual = cfop

File: apps/erpnext_fiscal_br/erpnext_fiscal_br/public/js/item.js
/**
 * Customizações para Item
 */

frappe.ui.form.on("Item", {
    ncm: function(frm) {
        // Remove formatação do NCM
        let ncm = frm.doc.ncm;
        if (!ncm) return;
        
        ncm = ncm.replace(/\D/g, "");
        
        if (ncm.length !== 8) {
            frappe.show_alert({
                message: __("NCM deve ter 8 dígitos"),
                indicator: "orange"
            });
        }
        
        frm.set_value("ncm", ncm);
        
        // Preenche gênero
        if (ncm.length >= 2) {
            frm.set_value("genero", ncm.substring(0, 2));
        }
    },
    
    cest: function(frm) {
        // Remove formatação do CEST
        let cest = frm.doc.cest;
        if (!cest) return;
        
        cest = cest.replace(/\D/g, "");
        
        if (cest.length !== 7) {
            frappe.show_alert({
                message: __("CEST deve ter 7 dígitos"),
                indicator: "orange"
            });
        }
        
        frm.set_value("cest", cest);
    }
});
